lib/question.js

```diff
@@ -0,0 +1,763 @@
+            sqldb.callWithClient(client, 'instance_questions_select_variant', params, (err, result) => {
+                if (ERR(err, callback)) return;
+                if (result.rowCount == 1) {
+                    const variant = result.rows[0];
+                    return callback(null, variant);
+                }
+                this._makeAndInsertVariantWithClient(client, question_id, instance_question_id, user_id, authn_user_id, course, options, (err, variant) => {
+                    if (ERR(err, callback)) return;
+                    callback(null, variant);
+                });
+            });
+        } else {
+            // if we don't have instance_question_id, just make a new variant
+            this._makeAndInsertVariantWithClient(client, question_id, instance_question_id, user_id, authn_user_id, course, options, (err, variant) => {
+                if (ERR(err, callback)) return;
+                callback(null, variant);
+            });
+        }
+    },
+
+    /**
+     * Ensure that there is a variant for the given instance question.
+     *
+     * @param {?number} question_id - The question for the new variant. Can be null if instance_question_id is provided.
+     * @param {?number} instance_question_id - The instance question for the new variant, or null for a floating variant.
+     * @param {number} user_id - The user for the new variant.
+     * @param {number} authn_user_id - The current authenticated user.
+     * @param {Object} course - The course for the variant.
+     * @param {Object} options - Options controlling the creation: options = {variant_seed}
+     * @param {boolean} require_open - If true, only use an existing variant if it is open.
+     * @param {function} callback - A callback(err, variant) function.
+     */
+    ensureVariant(question_id, instance_question_id, user_id, authn_user_id, course, options, require_open, callback) {
+        let variant;
+        sqldb.beginTransaction((err, client, done) => {
+            if (ERR(err, callback)) return;
+            async.series([
+                // Even though we only have a single series function,
+                // we use the async.series pattern for consistency and
+                // to make sure we correctly call endTransaction even
+                // in the presence of errors.
+                (callback) => {
+                    this._ensureVariantWithClient(client, question_id, instance_question_id, user_id, authn_user_id, course, options, require_open, (err, ret_variant) => {
+                        if (ERR(err, callback)) return;
+                        variant = ret_variant;
+                        callback(null);
+                    });
+                },
+            ], (err) => {
+                sqldb.endTransaction(client, done, err, (err) => {
+                    if (ERR(err, callback)) return;
+                    callback(null, variant);
+                });
+            });
+        });
+    },
+
+    /*
```

Comment:
Thanks! Running `make` in `doc/` will build the docs into `doc/code-docs`. Feel free to rearrange anything about this that you like. I was just trying to make some small progress towards better internal docs while refactoring.